app/components/steps/why.js

```diff
@@ -0,0 +1,244 @@
+    pointWhyList = [],
+    category = '', // "most" or "least"
+    onDone = () => {},
+    ...otherProps
+  } = props
+
+  const classes = useStylesFromThemeFunction()
+
+  if (!pointWhyList || pointWhyList.length === 0) {
+    return null // Return null if there is no data
+  }
+
+  // Initialize allWhys
+  const [allWhys, setAllWhys] = useState(() => {
+    const initialWhys = {}
+    pointWhyList.forEach(({ point, why }) => {
+      initialWhys[point._id] = why || {
+        subject: '',
+        description: '',
+        parentId: point._id,
+        category,
+      }
+    })
+
+    return initialWhys
+  })
+
+  // Sync changes in pointWhyList to allWhys
+  useEffect(() => {
+    const updatedWhys = {}
+    pointWhyList.forEach(({ point, why }) => {
+      updatedWhys[point._id] = {
+        ...allWhys[point._id],
+        ...(why || { subject: '', description: '', parentId: point._id, category }),
+      }
+    })
+    if (!isEqual(allWhys, updatedWhys)) {
+      setAllWhys(updatedWhys)
+    }
+  }, [pointWhyList])
+
+  const handleOnDone = ({ valid, value, delta }) => {
+    setAllWhys(prevWhys => {
+      const updatedWhys = {
+        ...prevWhys,
+        [delta.parentId]: { ...prevWhys[delta.parentId], ...delta },
+      }
+      const allValid = Object.values(updatedWhys).every(why => why.subject.trim() && why.description.trim())
+      onDone({ valid: allValid, value: allValid ? Object.values(updatedWhys) : value, delta })
+      return updatedWhys
+    })
+  }
+
+  return (
+    <div className={cx(classes.wrapper, className)} {...otherProps}>
+      <div className={classes.introContainer}>
+        <H className={classes.introTitle}>{`Why it's ${category && category[0].toUpperCase() + category.slice(1)} Important`}</H>
+        <div className={classes.introText}>{intro}</div>
+      </div>
+      <Level>
+        <div className={classes.pointsContainer}>
+          {pointWhyList.map(({ point }) => {
+            return (
+              <div key={point._id}>
+                <hr className={classes.pointsHr}></hr>
+                <WhyInput
+                  point={point}
+                  value={allWhys[point._id] || {}}
+                  onDone={({ valid, value }) => {
+                    const delta = {
+                      subject: value.subject || '',
+                      description: value.description || '',
+                      parentId: point._id,
+                      category,
+                    }
+                    handleOnDone({ valid, value: delta, delta })
+                  }}
+                />
+              </div>
+            )
+          })}
+        </div>
+      </Level>
+    </div>
+  )
+}
+
+export function userDeriver(data) {
+  const local = useRef({}).current
+  const reducedPointList = data?.reducedPointList || []
+  const myWhyByParentId = data?.myWhyByParentId || {}
+  const category = data?.category
+  const intro = data?.intro
+
+  const dataChanged = !isEqual(data, local.prevData)
```

Comment:
data is going to have a lot of other objects in it, from other step components, and isEqual is a lot of work to see if there are changes.

Also, what we are doing in these derivers, like React, is looking at the ref of an object to see if it is different from the previous ref.
`local.reducedPointList!==data.reducedPointList`
and then if anything needs to change deep in pointWhyList then we change it and also make sure that the ref of the pointWhyList  object changes.

If deepClone is used, everything changes.  But if there are 10 items in whyPointList and only one of them changes, then we want to change only that one, so that the other items in pointWhyList don't have to rerender.

I understand this is a challenging concept, look at the deriver in rerank.js

Also, would it help if I gave a talk on this in the next dev meeting?

I'm still reviewing everything else, but please give this a try.